src/stan_math_backend/Lower_functions.ml

```diff
@@ -204,141 +189,113 @@ let lower_fun_def (functors : (string, struct_defn) Hashtbl.t)
+    | {pattern= FunApp (StanLib (hof, FnPlain, _), args); _} ->
+        let f accum = function
+          | { pattern= Var name
+            ; meta= {Expr.Typed.Meta.type_= UnsizedType.UFun (args, _, _, _); _}
+            } ->
+              Map.add_multi accum
+                ~key:(Utils.stdlib_distribution_name name)
+                ~data:(Lower_expr.functor_type hof, List.map ~f:snd args)
+          | e -> find_functors_expr accum e in
+        List.fold ~init ~f args
+    | {pattern; _} -> Pattern.fold find_functors_expr init pattern in
   let rec find_functors_stmt accum stmt =
     Stmt.Fixed.(
       Pattern.fold find_functors_expr find_functors_stmt accum stmt.pattern)
   in
   Program.fold find_functors_expr find_functors_stmt String.Map.empty p
 
-(** We need to do a fair bit of bookkeeping to handle the functors necessary for the various
-    higher order functions.
-
-    Each functor needs a forward decl struct before the function,
-    then the function definition,
-    then the actual functor definitions
-  *)
 let collect_functors_functions (p : Program.Numbered.t) : defn list =
-  let (functors : (string, Cpp.struct_defn) Hashtbl.t) =
-    String.Table.create () in
-  let forward_decls = Hash_set.Poly.create () in
-  let reduce_sum_fns = is_fun_used_with_reduce_sum p in
-  let variadic_fns = get_variadic_requirements p in
-  let fun_and_functor_defs =
+  let functor_required = get_functor_requirements p in
+  let fun_has_def =
+    List.filter_map p.functions_block ~f:(function
+      | {fdname; fdbody= Some _; _} -> Some fdname
+      | _ -> None )
+    |> String.Set.of_list in
+  (* overloaded functions generate only one functor struct per name *)
+  let structs = String.Table.create () in
+  let is_overload (d : _ Program.fun_def) arg_types =
+    List.equal UnsizedType.equal
+      (List.map ~f:(fun (_, _, t) -> t) d.fdargs)
+      arg_types in
+  let declare_and_define (d : _ Program.fun_def) =
+    let functors =
+      Map.find_multi functor_required d.fdname
+      |> List.stable_dedup
+      |> List.filter_map ~f:(fun (hof, ts) ->
+             if is_overload d ts then Some hof else None ) in
+    let fn, st = lower_fun_def functors d in
+    List.iter st ~f:(fun s ->
+        (* Side effecting, collates functor structs *)
+        Hashtbl.update structs s.struct_name ~f:(function
+          | Some x -> {x with body= x.body @ s.body}
+          | None -> s ) ) ;
+    let decl, defn = Cpp.split_fun_decl_defn fn in
+    (FunDef decl, FunDef defn) in
+  let fun_decls, fun_defns =
     p.functions_block
-    |> List.concat_map
-         ~f:(lower_fun_def functors forward_decls reduce_sum_fns variadic_fns)
-    |> List.map ~f:(fun f -> FunDef f) in
-  let functor_struct_decls =
-    functors |> Hashtbl.data |> List.map ~f:(fun s -> Struct s) in
-  functor_struct_decls @ fun_and_functor_defs
+    |> List.filter_map ~f:(fun d ->
+           if Set.mem fun_has_def d.fdname && Option.is_none d.fdbody then None
```

Comment:
This can be falsely triggered by an overload, I think?

```stan
functions {
   void foo();

   void foo(int a) {
    print(a);
   }
}
```

Reply:
Oops, didn't think of that.
I think it's going to be easier to filter out the extra decls during AST -> MIR transformation and then not worry about it in the backend.